extension/src/lib.rs

```diff
@@ -0,0 +1,298 @@
+
+	// Prefix params with version, pallet, index to simplify decoding.
+	params.insert(0, version);
+	params.insert(1, pallet_index);
+	params.insert(2, call_index);
+	let (mut encoded_version, mut encoded_read) = (&params[..1], &params[1..]);
+	let version = decode_checked::<VersionedStateRead>(&mut encoded_version)?;
+
+	// Charge weight for doing one storage read.
+	env.charge_weight(T::DbWeight::get().reads(1_u64))?;
+	let result = match version {
+		VersionedStateRead::V0 => {
+			let read = decode_checked::<T::RuntimeRead>(&mut encoded_read)?;
+			ensure!(T::AllowedApiCalls::contains(&read), UNKNOWN_CALL_ERROR);
+			T::RuntimeRead::handle_read(read)
+		},
+	};
+	log::trace!(
+		target:LOG_TARGET,
+		"{} result: {:?}.", LOG_PREFIX, result
+	);
+	env.write(&result, false, None)
+}
+
+fn dispatch<T, E>(
+	env: &mut Environment<E, BufInBufOutState>,
+	version: u8,
+	pallet_index: u8,
+	call_index: u8,
+	mut params: Vec<u8>,
+) -> Result<(), DispatchError>
+where
+	T: Config,
+	E: Ext<T = T>,
+{
+	const LOG_PREFIX: &str = " dispatch |";
+
+	// Prefix params with version, pallet, index to simplify decoding.
+	params.insert(0, version);
+	params.insert(1, pallet_index);
+	params.insert(2, call_index);
+	let call = decode_checked::<VersionedDispatch<T>>(&mut &params[..])?;
+	// Contract is the origin by default.
+	let origin: T::RuntimeOrigin = RawOrigin::Signed(env.ext().address().clone()).into();
+	match call {
+		VersionedDispatch::V0(call) => dispatch_call::<T, E>(env, call, origin, LOG_PREFIX),
+	}
+}
+
+fn dispatch_call<T, E>(
+	env: &mut Environment<E, BufInBufOutState>,
+	call: T::RuntimeCall,
+	mut origin: T::RuntimeOrigin,
+	log_prefix: &str,
+) -> Result<(), DispatchError>
+where
+	T: Config,
+	E: Ext<T = T>,
+{
+	let charged_dispatch_weight = env.charge_weight(call.get_dispatch_info().weight)?;
+	log::debug!(target:LOG_TARGET, "{} Inputted RuntimeCall: {:?}", log_prefix, call);
+	origin.add_filter(T::AllowedApiCalls::contains);
+	match call.dispatch(origin) {
+		Ok(info) => {
+			log::debug!(target:LOG_TARGET, "{} success, actual weight: {:?}", log_prefix, info.actual_weight);
+			// Refund weight if the actual weight is less than the charged weight.
+			if let Some(actual_weight) = info.actual_weight {
+				env.adjust_weight(charged_dispatch_weight, actual_weight);
+			}
+			Ok(())
+		},
+		Err(err) => {
+			log::debug!(target:LOG_TARGET, "{} failed: error: {:?}", log_prefix, err.error);
+			Err(err.error)
+		},
+	}
+}
+
+/// Wrapper to enable versioning of runtime state reads.
+#[derive(Decode, Debug)]
+enum VersionedStateRead {
+	/// Version zero of state reads.
+	#[codec(index = 0)]
+	V0,
+}
+
+/// Wrapper to enable versioning of runtime calls.
+#[derive(Decode, Debug)]
+enum VersionedDispatch<T: Config> {
+	/// Version zero of dispatch calls.
+	#[codec(index = 0)]
+	V0(T::RuntimeCall),
+}
+
+// Converts a `DispatchError` to a `u32` status code based on the version of the API the contract uses.
```

Comment:
Could we move all this code and doc below the TryFrom implementation so that the types `Versioned...` and `FuncId` are nicely positioned together.